# Task
Add an introduction about the `diabetes.csv` dataset to the notebook as a text cell.

# Author/Developer: Navjyot Data Architect

## Add Dataset Introduction to Notebook

### Subtask:
Create a new text cell in the notebook containing the introduction about the `diabetes.csv` dataset, outlining its purpose, typical features, and target variable.


### Dataset Introduction: `diabetes.csv`

This dataset is commonly used for predicting the onset of diabetes based on various diagnostic measurements. It typically includes several medical predictor variables and one target variable.

**Purpose:** The primary goal is to predict whether a patient has diabetes based on diagnostic measurements included in the dataset.

**Typical Features (Predictor Variables):**
*   `Pregnancies`: Number of times pregnant
*   `Glucose`: Plasma glucose concentration a 2 hours in an oral glucose tolerance test
*   `BloodPressure`: Diastolic blood pressure (mm Hg)
*   `SkinThickness`: Triceps skin fold thickness (mm)
*   `Insulin`: 2-Hour serum insulin (mu U/ml)
*   `BMI`: Body mass index (weight in kg/(height in m)^2)
*   `DiabetesPedigreeFunction`: Diabetes pedigree function
*   `Age`: Age (years)

**Target Variable:**
*   `Outcome`: A binary variable (0 or 1), where 1 indicates the presence of diabetes and 0 indicates the absence of diabetes.

## Final Task

### Subtask:
Confirm that the dataset introduction has been successfully added to the notebook as a text cell.


## Summary:

### Q&A
The dataset introduction has been successfully added to the notebook as a text cell.

### Data Analysis Key Findings
- A markdown text cell was successfully generated and added to the notebook, providing an introduction to the `diabetes.csv` dataset.
- The introduction details the dataset's purpose: predicting diabetes onset based on various diagnostic measurements.
- It lists typical predictor variables such as `Pregnancies`, `Glucose`, `BloodPressure`, `SkinThickness`, `Insulin`, `BMI`, `DiabetesPedigreeFunction`, and `Age`.
- The target variable, `Outcome`, is clearly defined as a binary variable (0 for no diabetes, 1 for diabetes).

### Insights or Next Steps
- The foundational step of understanding the dataset through its introduction is complete, setting the stage for subsequent data analysis.
- The next logical step would be to load the dataset and perform an initial exploratory data analysis (EDA) to understand its structure and distributions.


## Data Understanding: Load the Dataset

### Subtask:
Load the `diabetes.csv` file into a pandas DataFrame. This is the essential first step to interact with our data.

In [1]:
import pandas as pd

# Load the dataset into a pandas DataFrame
df = pd.read_csv('/content/diabetes.csv')

print("Dataset loaded successfully!")

Dataset loaded successfully!


## Initial Data Inspection: Display Head and Tail

### Subtask:
Display the first 5 rows (`.head(5)`) and the last 5 rows (`.tail(5)`) of the DataFrame. This helps us quickly verify the data has been loaded correctly and to get a sense of its structure and content.

In [2]:
print("\nFirst 5 rows of the dataset:")
display(df.head(5))

print("\nLast 5 rows of the dataset:")
display(df.tail(5))


First 5 rows of the dataset:


Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1


Last 5 rows of the dataset:


Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0

# 📘 DATA UNDERSTANDING (DU)

Below are the exact steps included in the Data Understanding Phase:

---

## 1. Load the Dataset
- Use `pd.read_csv()`
- Display first 5 rows using `df.head()`

---

## 2. Dataset Shape
- `df.shape`
- Number of rows and columns

---

## 3. Column Names
- `df.columns`
- Check for spelling issues or extra spaces

---

## 4. Data Types
- `df.info()`
- Identify numeric, categorical, datetime, object types

---

## 5. Memory Usage
- `df.memory_usage(deep=True)`

---

## 6. Summary Statistics
- `df.describe()` for numeric
- `df.describe(include='object')` for categorical

---

## 7. Unique Values
- `df.nunique()`  
- Helps understand categorical columns

---

## 8. Missing Values Overview
- `df.isna().sum()`  
- Also calculate missing percentage

---

## 9. Duplicate Rows
- `df.duplicated().sum()`

---

## 10. Target Column Understanding
- Value counts
- Percentage distribution
- Is dataset balanced or imbalanced?

---

## 11. Domain Knowledge Notes
- Explanation of Glucose, BMI, Blood Pressure, Outcome

---

## 12. Identify Obvious Issues
Example impossible values:
- Glucose = 0  
- BloodPressure = 0  
- BMI = 0  

(Just identify here; fixing comes in Feature Engineering)

---

## 13. DU Summary
Include:
- Dataset characteristics  
- Issues found  
- What needs to be addressed in EDA

---



## All the code start from here above one is the explaination of data understanding and what are the steps in that I have mentioned in details and please follow all the steps which usally used in the industry standard.

## 2. Dataset Shape

### Subtask:
Determine the number of rows and columns in the DataFrame using `df.shape`. This provides an immediate understanding of the dataset's dimensions.

In [3]:
print("The dataset has {} rows and {} columns.".format(df.shape[0], df.shape[1]))

The dataset has 768 rows and 9 columns.


## 3. Column Names

### Subtask:
Retrieve and display the names of all columns in the DataFrame using `df.columns`. This step is crucial for verifying column labels and identifying any inconsistencies.

In [4]:
print("Column Names:\n", df.columns)

Column Names:
 Index(['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age', 'Outcome'],
      dtype='object')


## 4. Data Types

### Subtask:
Identify the data types of each column and check for non-null values using `df.info()`. This step is crucial for understanding the nature of the data and identifying potential issues like incorrect data types or missing values.

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


## 5. Memory Usage

### Subtask:
Calculate the memory usage of each column in the DataFrame using `df.memory_usage(deep=True)`. This helps in understanding the memory footprint of the dataset and individual columns.

In [6]:
print("\nMemory Usage of DataFrame (in bytes):")
display(df.memory_usage(deep=True))


Memory Usage of DataFrame (in bytes):


Index                        132
Pregnancies                 6144
Glucose                     6144
BloodPressure               6144
SkinThickness               6144
Insulin                     6144
BMI                         6144
DiabetesPedigreeFunction    6144
Age                         6144
Outcome                     6144
dtype: int64

## 6. Summary Statistics

### Subtask:
Generate descriptive statistics for the numerical columns using `df.describe()` and for categorical (object) columns using `df.describe(include='object')`. This provides insights into the central tendency, dispersion, and shape of the data's distribution, helping us understand key characteristics like mean, standard deviation, and quartiles.

In [10]:
print("\nDescriptive Statistics for Numerical Columns:")
display(df.describe())




Descriptive Statistics for Numerical Columns:


Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

## 7. Unique Values

### Subtask:
Calculate the number of unique values for each column in the DataFrame using `df.nunique()`. This helps in understanding the cardinality of each feature and is especially useful for identifying categorical columns or columns with limited variations.

In [11]:
print("\nNumber of Unique Values per Column:")
display(df.nunique())


Number of Unique Values per Column:


Pregnancies                  17
Glucose                     136
BloodPressure                47
SkinThickness                51
Insulin                     186
BMI                         248
DiabetesPedigreeFunction    517
Age                          52
Outcome                       2
dtype: int64

## 8. Missing Values Overview

### Subtask:
Calculate the total number of missing values for each column using `df.isna().sum()` and also compute the percentage of missing values. This step is essential for assessing data quality and planning for data imputation or removal strategies.

In [12]:
missing_values = df.isna().sum()
missing_percentage = (df.isna().sum() / len(df)) * 100

missing_info = pd.DataFrame({
    'Missing Values': missing_values,
    'Missing Percentage': missing_percentage
})

print("\nMissing Values Overview:")
display(missing_info[missing_info['Missing Values'] > 0].sort_values(by='Missing Values', ascending=False))

if missing_info['Missing Values'].sum() == 0:
    print("No missing values found in the dataset.")


Missing Values Overview:


Empty DataFrame
Columns: [Missing Values, Missing Percentage]
Index: []

No missing values found in the dataset.


## 9. Duplicate Rows

### Subtask:
Check for duplicate rows in the DataFrame using `df.duplicated().sum()`. This helps in identifying and quantifying any redundant entries in the dataset.

In [13]:
duplicate_rows = df.duplicated().sum()
print(f"Number of duplicate rows: {duplicate_rows}")

if duplicate_rows == 0:
    print("No duplicate rows found in the dataset.")

Number of duplicate rows: 0
No duplicate rows found in the dataset.


## 10. Target Column Understanding

### Subtask:
Analyze the distribution of the target variable (`Outcome`) by calculating its value counts and percentage distribution. This helps in determining if the dataset is balanced or imbalanced, which is important for model training considerations.

In [14]:
print("\nValue counts for the 'Outcome' column:")
display(df['Outcome'].value_counts())

print("\nPercentage distribution for the 'Outcome' column:")
display(df['Outcome'].value_counts(normalize=True) * 100)


Value counts for the 'Outcome' column:


Outcome
0    500
1    268
Name: count, dtype: int64


Percentage distribution for the 'Outcome' column:


Outcome
0    65.104167
1    34.895833
Name: proportion, dtype: float64

## 11. Domain Knowledge Notes

### Subtask:
Provide domain-specific explanations for key features like `Glucose`, `BMI`, `BloodPressure`, and `Outcome`. This helps in understanding the real-world implications of the data and identifying potential outliers or impossible values based on medical knowledge.

- **Glucose**: Typically, a fasting glucose level below 100 mg/dL is considered normal. Levels of 100-125 mg/dL indicate prediabetes, and 126 mg/dL or higher on two separate tests indicates diabetes. A value of 0 in this column is medically impossible for a living person and would likely represent a missing value.

- **BloodPressure**: Normal blood pressure is generally around 120/80 mmHg. Values much lower than 60 mmHg or higher than 140 mmHg (for diastolic, which is likely represented here) can indicate health issues. A value of 0 is medically impossible and suggests a missing value.

- **SkinThickness**: This measures the thickness of the skin at the triceps. While it varies, a value of 0 is biologically implausible for a living person and should be treated as a missing value.

- **Insulin**: Normal fasting insulin levels are typically between 2-25 mIU/L. Very low or very high values can indicate health problems. A value of 0 is medically impossible and likely indicates a missing value.

- **BMI**: Body Mass Index is calculated as weight (kg) / height (m)^2. A healthy BMI range is typically 18.5 to 24.9. A BMI of 0 is medically impossible for a living person and indicates a missing value.

- **Outcome**: This is our target variable, indicating the presence (1) or absence (0) of diabetes. It's a binary classification target.

## 12. Identify Obvious Issues

### Subtask:
Identify and quantify obvious impossible values (e.g., 0 for `Glucose`, `BloodPressure`, `BMI`, `SkinThickness`, `Insulin`) as these are medically implausible and likely represent missing data. (Just identify here; fixing comes in Feature Engineering)

In [15]:
print("\nNumber of 0 values in key features (indicating impossible values):\n")

# List of columns to check for 0 values that are medically implausible
columns_with_impossible_zeros = ['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']

for col in columns_with_impossible_zeros:
    zero_count = (df[col] == 0).sum()
    if zero_count > 0:
        print(f"{col}: {zero_count} ({(zero_count / len(df)) * 100:.2f}%)")
    else:
        print(f"{col}: {zero_count}")



Number of 0 values in key features (indicating impossible values):

Glucose: 5 (0.65%)
BloodPressure: 35 (4.56%)
SkinThickness: 227 (29.56%)
Insulin: 374 (48.70%)
BMI: 11 (1.43%)


## 13. DU Summary

### Subtask:
Summarize the key findings from the Data Understanding phase, including dataset characteristics, issues found, and what needs to be addressed in the Exploratory Data Analysis (EDA) phase.

### Data Understanding Summary:

**Dataset Characteristics:**
- **Shape:** The dataset contains 768 rows and 9 columns.
- **Column Names:** All column names are standard and well-defined (e.g., `Pregnancies`, `Glucose`, `BloodPressure`, `SkinThickness`, `Insulin`, `BMI`, `DiabetesPedigreeFunction`, `Age`, `Outcome`).
- **Data Types:** All columns are numerical, with 7 `int64` and 2 `float64` types. There are no categorical columns.
- **Memory Usage:** The DataFrame has a memory footprint of approximately 54.1 KB.
- **Target Variable (`Outcome`):** This is a binary classification target with 2 unique values (0 and 1). The distribution is imbalanced, with approximately 65.10% indicating no diabetes (0) and 34.90% indicating diabetes (1).

**Issues Found:**
- **No explicit missing values:** `df.isna().sum()` reported no missing values.
- **No duplicate rows:** `df.duplicated().sum()` reported no duplicate rows.
- **Implausible Zero Values:** A significant issue identified is the presence of 0 values in several key medical measurement columns (`Glucose`, `BloodPressure`, `SkinThickness`, `Insulin`, `BMI`). These values are medically impossible for living individuals and should be treated as missing data:
    - `Glucose`: 5 instances (0.65%)
    - `BloodPressure`: 35 instances (4.56%)
    - `SkinThickness`: 227 instances (29.56%)
    - `Insulin`: 374 instances (48.70%)
    - `BMI`: 11 instances (1.43%)

**What Needs to be Addressed in EDA (and subsequent Feature Engineering):**
- **Handling Implausible Zero Values:** The 0 values in `Glucose`, `BloodPressure`, `SkinThickness`, `Insulin`, and `BMI` must be appropriately handled. These will likely be imputed or replaced, as they do not represent true measurements. This will be a critical step in the Feature Engineering phase.
- **Class Imbalance:** The target variable `Outcome` shows a class imbalance. This should be considered during model training to avoid biased predictions, potentially requiring techniques like oversampling, undersampling, or using appropriate evaluation metrics.
- **Outlier Detection:** While `df.describe()` showed ranges, further visual exploration (e.g., box plots, histograms) during EDA will be crucial to identify and understand outliers in numerical features, especially given the presence of 0s that are not true values.

This concludes our Data Understanding phase. We now have a solid foundation to move into Exploratory Data Analysis!